lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -345,12 +345,14 @@ static bool isInsideCompatibleUnavailableDeclaration(
     return false;
   }
 
-  // Refuse calling unavailable functions from unavailable code,
-  // but allow the use of types.
+  // Unless in embedded Swift mode, refuse calling unavailable functions from
+  // unavailable code, but allow the use of types.
   PlatformKind platform = attr->Platform;
-  if (platform == PlatformKind::none && !isa<TypeDecl>(D) &&
-      !isa<ExtensionDecl>(D)) {
-    return false;
+  if (!D->getASTContext().LangOpts.hasFeature(Feature::Embedded)) {
+    if (platform == PlatformKind::none && !isa<TypeDecl>(D) &&
+        !isa<ExtensionDecl>(D)) {
```

Comment:
I don't fully understand the logic here. This is just to get things to typecheck when embedded is not enabled? Why do we need to check the platform (or decl kind for that matter)?

Reply:
As the availability checking logic currently stands, this is invalid:
```
@available(*, unavailable)
func foo() {}

@available(*, unavailable)
func bar() {
  foo() // error
}
```

However, if `@available(*, unavailable)` were instead `@available(macOS, unavailable)` it would type check successfully. I think the rationale is that universally unavailable functions are never callable anywhere, so they should not be allowed anywhere. However, I think we should probably generally make the exception that calls to universally unavailable functions are allowed from other universally unavailable contexts. I can follow up later to make this more general though.